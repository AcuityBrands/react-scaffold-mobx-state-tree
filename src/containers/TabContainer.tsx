/**
 * TabNavigation Component
 * 
 * Primary content render engine for application.  This component
 * Will listen to changes to the browser location and render
 * a new tab (or reuse an existing one).
 * 
 * WARNING: Dependent on DOM router
 */
import * as React from "react";
import { inject } from "mobx-react";
import { withRouter } from 'react-router-dom';
import { Tabs } from 'antd';
import { IAppStore } from '../stores/AppStore';
import { LazyDashboard } from '../containers/LazyDashboard'
import { LazyAbout } from '../containers/LazyAbout'
import { LazyTodo } from '../containers/LazyTodo'
const TabPane = Tabs.TabPane;

interface ITabNavProps {
  match: any,
  location: any,
  history: any,
  appStore: IAppStore
}

interface IPane {
  title: string,
  content: any,
  key: string,
  disabled: boolean
}

interface ITabState {
  activeKey: string,
  panes: Array<IPane>
}


@inject("appStore")
class TabContainer extends React.Component<ITabNavProps, {}>{
  state: ITabState
  unlisten: any;

  // TODO: Move route configurations to common file for reusability
  pathConfig = [
    { path: "/dashboard", label: "Dashboard", component: <LazyDashboard/>},
    { path: "/task", label: "Tasks", component: <LazyTodo/>},
    { path: "/about", label: "About App", component: <LazyAbout/>}
  ]

  constructor(props: any) {
    super(props);
    this.state = {
      activeKey: '',
      panes: []
    };
  }

  componentDidMount() {
    this.renderComponent(this.props.location);
    this.unlisten = this.props.history.listen((location: any, action: any) => {
      this.renderComponent(location);
    });
  }

  onEdit = (targetKey: string, action: string) => {
    if (action === 'remove') this.remove(targetKey);
  }

  onChange = (activeKey: string) => {
    this.setState({ activeKey: activeKey });
    this.props.history.push(activeKey);
  }

  renderComponent = (location: any) => {
    const panes = this.state.panes;

    // Determine if tab is already open
    // if so, set the active tab to the open tab
    for (let x = 0; x < panes.length; x++) {
      let p = panes[x] as IPane;
      if (p.key === location.pathname) {
        this.setState({ activeKey: p.key });
        return false;
      }
    }

    // Lookup the pane content given the
    // router location
    let item;
    let pathFound: boolean = false;
    for (let i = 0; i < this.pathConfig.length; i++) {
      item = this.pathConfig[i];
      if (item.path === location.pathname) {
        pathFound = true;
        break;
      }
    }

    // If the path is found, create a new tab
    if (pathFound) {
      panes.push({ title: item.label, content: item.component, key: item.path, disabled: true });
      if (panes.length > 1) {
        panes.map(p => p.disabled = false);
      }
      this.setState({ panes, activeKey: item.path });
    }

  }

  remove = (targetKey: string) => {
    let keyIndex: number = -1;
    let activeKey = this.state.activeKey;

    // Find the index of the deleted key
    for (let i = 0; i < this.state.panes.length; i++) {
      let p = this.state.panes[i];
      if (p.key == targetKey) {
        keyIndex = i;
        break;
      }
    }

    // Remove the deleted key from the panes
    const panes = this.state.panes.filter(e => e.key != targetKey);

    // If tab removed was the first tab
    if (keyIndex == 0 && targetKey == activeKey && panes.length > 0)
      activeKey = panes[0].key;

    // If tab removed was the last tab
    if (keyIndex == panes.length && panes.length > 0)
      activeKey = panes[panes.length - 1].key

    // Disable tabs if only 1 tab remains
    if (panes.length <= 1) {
      panes.map(p => p.disabled = true);
    }

    // Update State
    this.setState({ panes, activeKey });
    this.props.history.push(activeKey);
  }

  render() {
    return (
      <div className="tab-wrapper">
        <Tabs
          hideAdd
          activeKey={this.state.activeKey}
          type="editable-card"
          onEdit={this.onEdit}
          onChange={this.onChange}
          className="custom-tabs"
        >
          {this.state.panes.map((pane: IPane) => <TabPane disabled={pane.disabled} tab={pane.title} key={pane.key}>{pane.content}</TabPane>)}
        </Tabs>
      </div>
    );
  }
}

export default withRouter(TabContainer);