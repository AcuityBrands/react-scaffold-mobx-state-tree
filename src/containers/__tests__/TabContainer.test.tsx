import * as React from "react";
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { TabContainer } from '../TabContainer';

Enzyme.configure({ adapter: new Adapter() });

const setup = propOverrides => {
  const props = Object.assign({
    config: [
      { key: "1", path: "/dashboard", icon: "line-chart", label: "Dashboard", component: "Dashboard"},
      { key: "2", path: "/task", icon: "check-circle-o", label: "Tasks", component: "Todo"},
      { key: "3", component: "Divider"},
      { key: "4", path: "/about", icon:"info-circle-o", label: "About App", component: "About"}
    ],
    location: {
      pathname: "/dashboard"
    },
    history: {
      listen: jest.fn(),
      push: jest.fn()
    }
  }, propOverrides)
  const wrapper = Enzyme.shallow(<TabContainer {...props} />);
  return {
    props,
    wrapper,
    tabs: wrapper.find("Tabs")
  }
}

describe('TabContainer Component', () => {

  it("can render", () => {
    const { wrapper } = setup({});
    expect(wrapper.exists()).toBe(true)
  })

  it("matches the snapshot", () => {
    const { wrapper } = setup({});
    expect(wrapper).toMatchSnapshot();
  })

  it("contains a root Tab component", () => {
    const { wrapper, tabs } = setup({});
    expect(tabs.length).toBe(1);
  })

  it("will create a pane if location matches config", () => {
    const { wrapper, props, tabs } = setup({});
    expect(tabs.children().length).toBe(1);
  })

  it("will not create a pane if location does not match config", () => {
    const { wrapper, props, tabs } = setup({location:{pathname:"/garbage"}});
    expect(tabs.children().length).toBe(0);
  })

  it("remove a pane if removed via click", () => {
    const { wrapper, props, tabs } = setup({});
    tabs.simulate('edit', "/dashboard", "remove");
    // Calling these in this specific order to force a re-render
    // https://github.com/airbnb/enzyme/issues/1229
    wrapper.instance().forceUpdate();
    wrapper.update();
    // =========================================
    expect(wrapper.find("Tabs").children().length).toBe(0);
  })
})