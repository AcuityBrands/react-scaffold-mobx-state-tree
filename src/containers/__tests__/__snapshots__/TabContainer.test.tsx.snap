// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabContainer Component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TabContainer
    config={
      Array [
        Object {
          "component": "Dashboard",
          "icon": "line-chart",
          "key": "1",
          "label": "Dashboard",
          "path": "/dashboard",
        },
        Object {
          "component": "Todo",
          "icon": "check-circle-o",
          "key": "2",
          "label": "Tasks",
          "path": "/task",
        },
        Object {
          "component": "Divider",
          "key": "3",
        },
        Object {
          "component": "About",
          "icon": "info-circle-o",
          "key": "4",
          "label": "About App",
          "path": "/about",
        },
      ]
    }
    history={
      Object {
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
        "push": [MockFunction],
      }
    }
    location={
      Object {
        "pathname": "/dashboard",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Tabs
        activeKey="/dashboard"
        className="custom-tabs"
        hideAdd={true}
        onChange={[Function]}
        onEdit={[Function]}
        prefixCls="ant-tabs"
        type="editable-card"
      >
        <TabPane
          disabled={true}
          placeholder={null}
          tab="Dashboard"
        >
          Dashboard
        </TabPane>
      </Tabs>,
      "className": "tab-wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeKey": "/dashboard",
        "children": Array [
          <TabPane
            disabled={true}
            placeholder={null}
            tab="Dashboard"
          >
            Dashboard
          </TabPane>,
        ],
        "className": "custom-tabs",
        "hideAdd": true,
        "onChange": [Function],
        "onEdit": [Function],
        "prefixCls": "ant-tabs",
        "type": "editable-card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "/dashboard",
          "nodeType": "class",
          "props": Object {
            "children": "Dashboard",
            "disabled": true,
            "placeholder": null,
            "tab": "Dashboard",
          },
          "ref": null,
          "rendered": "Dashboard",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Tabs
          activeKey="/dashboard"
          className="custom-tabs"
          hideAdd={true}
          onChange={[Function]}
          onEdit={[Function]}
          prefixCls="ant-tabs"
          type="editable-card"
        >
          <TabPane
            disabled={true}
            placeholder={null}
            tab="Dashboard"
          >
            Dashboard
          </TabPane>
        </Tabs>,
        "className": "tab-wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeKey": "/dashboard",
          "children": Array [
            <TabPane
              disabled={true}
              placeholder={null}
              tab="Dashboard"
            >
              Dashboard
            </TabPane>,
          ],
          "className": "custom-tabs",
          "hideAdd": true,
          "onChange": [Function],
          "onEdit": [Function],
          "prefixCls": "ant-tabs",
          "type": "editable-card",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "/dashboard",
            "nodeType": "class",
            "props": Object {
              "children": "Dashboard",
              "disabled": true,
              "placeholder": null,
              "tab": "Dashboard",
            },
            "ref": null,
            "rendered": "Dashboard",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
